File: glp_runtime/lib/link/reliability/frame_codec.dart
import 'dart:typed_data';

import 'crc32.dart';
import 'frame_exception.dart';

/// The on-wire frame kind.
enum FrameKind {
  /// A whole payload in one frame (fits under MTU).
  whole, // 0

  /// One fragment of a payload split across several frames (FR-022).
  fragment, // 1
}

/// A parsed, CRC-validated frame: its header fields plus its chunk bytes.
///
/// [kind]: Whole or Fragment.
/// [messageId]: Groups the fragments of one logical payload.
/// [totalLength]: Full payload byte length (identical across a message's fragments).
/// [fragIndex]: 0-based fragment position (0 for [FrameKind.whole]).
/// [fragCount]: Number of fragments (1 for [FrameKind.whole]).
/// [chunk]: This frame's payload slice (already CRC-verified).
class ParsedFrame {
  final FrameKind kind;
  final int messageId;
  final int totalLength;
  final int fragIndex;
  final int fragCount;
  final Uint8List chunk;

  const ParsedFrame(this.kind, this.messageId, this.totalLength, this.fragIndex,
      this.fragCount, this.chunk);

  @override
  bool operator ==(Object other) =>
      other is ParsedFrame &&
      kind == other.kind &&
      messageId == other.messageId &&
      totalLength == other.totalLength &&
      fragIndex == other.fragIndex &&
      fragCount == other.fragCount &&
      _bytesEqual(chunk, other.chunk);

  @override
  int get hashCode => Object.hash(
      kind, messageId, totalLength, fragIndex, fragCount, Object.hashAll(chunk));

  static bool _bytesEqual(Uint8List a, Uint8List b) {
    if (a.length != b.length) return false;
    for (int i = 0; i < a.length; i++) {
      if (a[i] != b[i]) return false;
    }
    return true;
  }
}

/// The wire-format frame codec (T021, FR-022). Wraps the opaque
/// `PayloadSerializer` blob with what corpus 13 §6 flags as ABSENT today: a
/// version byte (forward-compat), an outer length + CRC-32 (the Dart path relied
/// on in-process object integrity), and fragmentation/reassembly for under-MTU
/// leaves (CoAP ~1 KB, BLE GATT 20 B). Big-endian throughout to match the
/// serializer's endian-independent TLV and so the Dart mirror is byte-identical
/// (FR-060/061).
///
/// The `messageId` is supplied by the caller (the link's outbound counter,
/// related to the T022 sequence number), NOT generated here — encoding is
/// therefore deterministic and replayable for the seeded loopback/parity tests.
class FrameCodec {
  FrameCodec._();

  /// Current wire version. A frame with any other version byte is rejected.
  static const int version = 0x01;

  /// Fixed header size in bytes (see field layout in [encode]).
  static const int headerSize = 22;

  /// Upper bound on a single payload (FR-028 deserializer hardening): a frame
  /// advertising a larger `totalLength` is rejected before any allocation,
  /// so a forged length cannot exhaust memory.
  static const int maxPayloadBytes = 64 * 1024 * 1024;

  // Header field offsets (big-endian):
  //   0      Version   (1)
  //   1      Kind      (1)
  //   2..5   MessageId (u32)
  //   6..9   TotalLen  (u32)
  //   10..11 FragIndex (u16)
  //   12..13 FragCount (u16)
  //   14..17 ChunkCrc  (u32)  CRC-32 of the chunk bytes
  //   18..21 ChunkLen  (u32)
  //   22..   chunk
  static const int _offKind = 1;
  static const int _offMessageId = 2;
  static const int _offTotalLen = 6;
  static const int _offFragIndex = 10;
  static const int _offFragCount = 12;
  static const int _offChunkCrc = 14;
  static const int _offChunkLen = 18;

  /// Encode [payload] into one or more self-delimiting frames.
  /// A single [FrameKind.whole] frame when it fits under
  /// [maxFrameBytes] (or when that is `null`), otherwise
  /// [FrameKind.fragment] frames each ≤ the MTU.
  ///
  /// [payload]: The opaque serialized payload.
  /// [messageId]: Caller-supplied id grouping this payload's frames.
  /// [maxFrameBytes]: MTU including header, or `null` for unbounded.
  static List<Uint8List> encode(List<int> payload, int messageId,
      [int? maxFrameBytes]) {
    if (payload.length > maxPayloadBytes) {
      throw FrameException(
          'payload ${payload.length} exceeds MaxPayloadBytes $maxPayloadBytes');
    }

    final int total = payload.length;

    final bool fitsWhole =
        maxFrameBytes == null || total + headerSize <= maxFrameBytes;
    if (fitsWhole) {
      return [
        _buildFrame(FrameKind.whole, messageId, total, 0, 1, payload, 0, total)
      ];
    }

    final int maxChunk = maxFrameBytes - headerSize;
    if (maxChunk <= 0) {
      throw FrameException(
          'maxFrameBytes $maxFrameBytes too small for $headerSize-byte header');
    }

    final int fragCountL =
        total == 0 ? 1 : (total + maxChunk - 1) ~/ maxChunk;
    if (fragCountL > 0xFFFF) {
      throw FrameException(
          'payload $total needs $fragCountL fragments under MTU $maxFrameBytes (max ${0xFFFF})');
    }
    final int fragCount = fragCountL;

    final frames = <Uint8List>[];
    for (int i = 0; i < fragCount; i++) {
      final int start = i * maxChunk;
      final int len = maxChunk < total - start ? maxChunk : total - start;
      frames.add(_buildFrame(FrameKind.fragment, messageId, total, i, fragCount,
          payload, start, len));
    }
    return frames;
  }

  static Uint8List _buildFrame(FrameKind kind, int messageId, int totalLen,
      int fragIndex, int fragCount, List<int> source, int chunkStart,
      int chunkLength) {
    final frame = Uint8List(headerSize + chunkLength);
    final h = ByteData.sublistView(frame);
    frame[0] = version;
    frame[_offKind] = kind.index;
    h.setUint32(_offMessageId, messageId & 0xFFFFFFFF, Endian.big);
    h.setUint32(_offTotalLen, totalLen & 0xFFFFFFFF, Endian.big);
    h.setUint16(_offFragIndex, fragIndex & 0xFFFF, Endian.big);
    h.setUint16(_offFragCount, fragCount & 0xFFFF, Endian.big);
    final chunkView = source is Uint8List
        ? Uint8List.sublistView(source, chunkStart, chunkStart + chunkLength)
        : Uint8List.fromList(
            source.sublist(chunkStart, chunkStart + chunkLength));
    h.setUint32(_offChunkCrc, Crc32.compute(chunkView), Endian.big);
    h.setUint32(_offChunkLen, chunkLength & 0xFFFFFFFF, Endian.big);
    frame.setRange(headerSize, headerSize + chunkLength, chunkView);
    return frame;
  }

  /// Parse and validate one frame: version byte, CRC-32 over the chunk, and
  /// header/length consistency. Throws [FrameException] on any
  /// mismatch (FR-022 bad-version/bad-CRC rejected) — never a silent mis-decode.
  static ParsedFrame parseFrame(Uint8List frame) {
    if (frame.length < headerSize) {
      throw FrameException(
          'frame ${frame.length} shorter than $headerSize-byte header');
    }

    final int versionByte = frame[0];
    if (versionByte != version) {
      throw FrameException(
          'unsupported wire version 0x${_hex2(versionByte)} (expected 0x${_hex2(version)})');
    }

    final int kindByte = frame[_offKind];
    if (kindByte != FrameKind.whole.index &&
        kindByte != FrameKind.fragment.index) {
      throw FrameException('unknown frame kind $kindByte');
    }
    final kind = FrameKind.values[kindByte];

    final h = ByteData.sublistView(frame);
    final int messageId = h.getUint32(_offMessageId, Endian.big);
    final int totalLen = h.getUint32(_offTotalLen, Endian.big);
    final int fragIndex = h.getUint16(_offFragIndex, Endian.big);
    final int fragCount = h.getUint16(_offFragCount, Endian.big);
    final int chunkCrc = h.getUint32(_offChunkCrc, Endian.big);
    final int chunkLen = h.getUint32(_offChunkLen, Endian.big);

    if (totalLen > maxPayloadBytes) {
      throw FrameException(
          'advertised total length $totalLen exceeds MaxPayloadBytes $maxPayloadBytes');
    }
    if (fragCount == 0) {
      throw FrameException('fragment count 0');
    }
    if (fragIndex >= fragCount) {
      throw FrameException(
          'fragment index $fragIndex out of range for count $fragCount');
    }
    if (kind == FrameKind.whole && fragCount != 1) {
      throw FrameException('whole frame with fragment count $fragCount');
    }

    final int expectedFrameLen = headerSize + chunkLen;
    if (frame.length != expectedFrameLen) {
      throw FrameException(
          'frame length ${frame.length} != header+chunk $expectedFrameLen');
    }

    final chunk =
        Uint8List.fromList(frame.sublist(headerSize, headerSize + chunkLen));
    final int actualCrc = Crc32.compute(chunk);
    if (actualCrc != chunkCrc) {
      throw FrameException(
          'CRC mismatch: got 0x${_hex8(actualCrc)}, frame says 0x${_hex8(chunkCrc)}');
    }

    return ParsedFrame(kind, messageId, totalLen, fragIndex, fragCount, chunk);
  }

  // Mirrors C# "{value:X2}" — upper-case hex, min width 2, zero-padded.
  static String _hex2(int value) =>
      value.toRadixString(16).toUpperCase().padLeft(2, '0');

  // Mirrors C# "{value:X8}" — upper-case hex, min width 8, zero-padded.
  static String _hex8(int value) =>
      value.toRadixString(16).toUpperCase().padLeft(8, '0');
}

File: csharp/glp_link/reliability/FrameCodec.cs
using System.Buffers.Binary;

namespace GlpRuntime.Link.Reliability;

/// <summary>The on-wire frame kind.</summary>
public enum FrameKind : byte
{
    /// <summary>A whole payload in one frame (fits under MTU).</summary>
    Whole = 0,

    /// <summary>One fragment of a payload split across several frames (FR-022).</summary>
    Fragment = 1,
}

/// <summary>A parsed, CRC-validated frame: its header fields plus its chunk bytes.</summary>
/// <param name="Kind">Whole or Fragment.</param>
/// <param name="MessageId">Groups the fragments of one logical payload.</param>
/// <param name="TotalLength">Full payload byte length (identical across a message's fragments).</param>
/// <param name="FragIndex">0-based fragment position (0 for <see cref="FrameKind.Whole"/>).</param>
/// <param name="FragCount">Number of fragments (1 for <see cref="FrameKind.Whole"/>).</param>
/// <param name="Chunk">This frame's payload slice (already CRC-verified).</param>
public readonly record struct ParsedFrame(
    FrameKind Kind, uint MessageId, uint TotalLength, ushort FragIndex, ushort FragCount, byte[] Chunk);

/// <summary>
/// The wire-format frame codec (T021, FR-022). Wraps the opaque
/// <c>PayloadSerializer</c> blob with what corpus 13 §6 flags as ABSENT today: a
/// version byte (forward-compat), an outer length + CRC-32 (the Dart path relied
/// on in-process object integrity), and fragmentation/reassembly for under-MTU
/// leaves (CoAP ~1 KB, BLE GATT 20 B). Big-endian throughout to match the
/// serializer's endian-independent TLV and so the Dart mirror is byte-identical
/// (FR-060/061).
/// </summary>
/// <remarks>
/// The <c>MessageId</c> is supplied by the caller (the link's outbound counter,
/// related to the T022 sequence number), NOT generated here — encoding is
/// therefore deterministic and replayable for the seeded loopback/parity tests.
/// </remarks>
public static class FrameCodec
{
    /// <summary>Current wire version. A frame with any other version byte is rejected.</summary>
    public const byte Version = 0x01;

    /// <summary>Fixed header size in bytes (see field layout in <see cref="Encode"/>).</summary>
    public const int HeaderSize = 22;

    /// <summary>
    /// Upper bound on a single payload (FR-028 deserializer hardening): a frame
    /// advertising a larger <c>TotalLength</c> is rejected before any allocation,
    /// so a forged length cannot exhaust memory.
    /// </summary>
    public const int MaxPayloadBytes = 64 * 1024 * 1024;

    // Header field offsets (big-endian):
    //   0      Version   (1)
    //   1      Kind      (1)
    //   2..5   MessageId (u32)
    //   6..9   TotalLen  (u32)
    //   10..11 FragIndex (u16)
    //   12..13 FragCount (u16)
    //   14..17 ChunkCrc  (u32)  CRC-32 of the chunk bytes
    //   18..21 ChunkLen  (u32)
    //   22..   chunk
    private const int OffKind = 1;
    private const int OffMessageId = 2;
    private const int OffTotalLen = 6;
    private const int OffFragIndex = 10;
    private const int OffFragCount = 12;
    private const int OffChunkCrc = 14;
    private const int OffChunkLen = 18;

    /// <summary>
    /// Encode <paramref name="payload"/> into one or more self-delimiting frames.
    /// A single <see cref="FrameKind.Whole"/> frame when it fits under
    /// <paramref name="maxFrameBytes"/> (or when that is <c>null</c>), otherwise
    /// <see cref="FrameKind.Fragment"/> frames each ≤ the MTU.
    /// </summary>
    /// <param name="payload">The opaque serialized payload.</param>
    /// <param name="messageId">Caller-supplied id grouping this payload's frames.</param>
    /// <param name="maxFrameBytes">MTU including header, or <c>null</c> for unbounded.</param>
    public static IReadOnlyList<byte[]> Encode(ReadOnlySpan<byte> payload, uint messageId, int? maxFrameBytes = null)
    {
        if (payload.Length > MaxPayloadBytes)
            throw new FrameException($"payload {payload.Length} exceeds MaxPayloadBytes {MaxPayloadBytes}");

        int total = payload.Length;

        bool fitsWhole = maxFrameBytes is null || total + HeaderSize <= maxFrameBytes;
        if (fitsWhole)
            return new[] { BuildFrame(FrameKind.Whole, messageId, (uint)total, 0, 1, payload) };

        int maxChunk = maxFrameBytes!.Value - HeaderSize;
        if (maxChunk <= 0)
            throw new FrameException($"maxFrameBytes {maxFrameBytes} too small for {HeaderSize}-byte header");

        long fragCountL = total == 0 ? 1 : (total + maxChunk - 1L) / maxChunk;
        if (fragCountL > ushort.MaxValue)
            throw new FrameException($"payload {total} needs {fragCountL} fragments under MTU {maxFrameBytes} (max {ushort.MaxValue})");
        ushort fragCount = (ushort)fragCountL;

        var frames = new List<byte[]>(fragCount);
        for (ushort i = 0; i < fragCount; i++)
        {
            int start = i * maxChunk;
            int len = Math.Min(maxChunk, total - start);
            frames.Add(BuildFrame(FrameKind.Fragment, messageId, (uint)total, i, fragCount, payload.Slice(start, len)));
        }
        return frames;
    }

    private static byte[] BuildFrame(FrameKind kind, uint messageId, uint totalLen, ushort fragIndex, ushort fragCount, ReadOnlySpan<byte> chunk)
    {
        var frame = new byte[HeaderSize + chunk.Length];
        var h = frame.AsSpan();
        h[0] = Version;
        h[OffKind] = (byte)kind;
        BinaryPrimitives.WriteUInt32BigEndian(h.Slice(OffMessageId), messageId);
        BinaryPrimitives.WriteUInt32BigEndian(h.Slice(OffTotalLen), totalLen);
        BinaryPrimitives.WriteUInt16BigEndian(h.Slice(OffFragIndex), fragIndex);
        BinaryPrimitives.WriteUInt16BigEndian(h.Slice(OffFragCount), fragCount);
        BinaryPrimitives.WriteUInt32BigEndian(h.Slice(OffChunkCrc), Crc32.Compute(chunk));
        BinaryPrimitives.WriteUInt32BigEndian(h.Slice(OffChunkLen), (uint)chunk.Length);
        chunk.CopyTo(h.Slice(HeaderSize));
        return frame;
    }

    /// <summary>
    /// Parse and validate one frame: version byte, CRC-32 over the chunk, and
    /// header/length consistency. Throws <see cref="FrameException"/> on any
    /// mismatch (FR-022 bad-version/bad-CRC rejected) — never a silent mis-decode.
    /// </summary>
    public static ParsedFrame ParseFrame(ReadOnlySpan<byte> frame)
    {
        if (frame.Length < HeaderSize)
            throw new FrameException($"frame {frame.Length} shorter than {HeaderSize}-byte header");

        byte version = frame[0];
        if (version != Version)
            throw new FrameException($"unsupported wire version 0x{version:X2} (expected 0x{Version:X2})");

        byte kindByte = frame[OffKind];
        if (kindByte is not ((byte)FrameKind.Whole or (byte)FrameKind.Fragment))
            throw new FrameException($"unknown frame kind {kindByte}");
        var kind = (FrameKind)kindByte;

        uint messageId = BinaryPrimitives.ReadUInt32BigEndian(frame.Slice(OffMessageId));
        uint totalLen = BinaryPrimitives.ReadUInt32BigEndian(frame.Slice(OffTotalLen));
        ushort fragIndex = BinaryPrimitives.ReadUInt16BigEndian(frame.Slice(OffFragIndex));
        ushort fragCount = BinaryPrimitives.ReadUInt16BigEndian(frame.Slice(OffFragCount));
        uint chunkCrc = BinaryPrimitives.ReadUInt32BigEndian(frame.Slice(OffChunkCrc));
        uint chunkLen = BinaryPrimitives.ReadUInt32BigEndian(frame.Slice(OffChunkLen));

        if (totalLen > MaxPayloadBytes)
            throw new FrameException($"advertised total length {totalLen} exceeds MaxPayloadBytes {MaxPayloadBytes}");
        if (fragCount == 0)
            throw new FrameException("fragment count 0");
        if (fragIndex >= fragCount)
            throw new FrameException($"fragment index {fragIndex} out of range for count {fragCount}");
        if (kind == FrameKind.Whole && fragCount != 1)
            throw new FrameException($"whole frame with fragment count {fragCount}");

        long expectedFrameLen = HeaderSize + (long)chunkLen;
        if (frame.Length != expectedFrameLen)
            throw new FrameException($"frame length {frame.Length} != header+chunk {expectedFrameLen}");

        var chunk = frame.Slice(HeaderSize, (int)chunkLen).ToArray();
        uint actualCrc = Crc32.Compute(chunk);
        if (actualCrc != chunkCrc)
            throw new FrameException($"CRC mismatch: got 0x{actualCrc:X8}, frame says 0x{chunkCrc:X8}");

        return new ParsedFrame(kind, messageId, totalLen, fragIndex, fragCount, chunk);
    }
}
